cdc/kv/region_state.go

```diff
@@ -0,0 +1,283 @@
+	return s.sri.verID.GetID()
+}
+
+func (s *regionFeedState) getRequestID() uint64 {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.requestID
+}
+
+func (s *regionFeedState) getCheckpointTs() uint64 {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.sri.checkpointTs
+}
+
+func (s *regionFeedState) getStoreAddr() string {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.sri.rpcCtx.Addr
+}
+
+func (s *regionFeedState) updateCheckpoint() {
+	s.lock.RLock()
+	resolvedTs := s.lastResolvedTs
+	checkpointTs := s.sri.checkpointTs
+	if resolvedTs <= checkpointTs {
+		s.lock.RUnlock()
+		return
+	}
+	s.lock.RUnlock()
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.sri.checkpointTs = resolvedTs
+}
+
+func (s *regionFeedState) getRegionInfo() singleRegionInfo {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.sri
+}
+
+func (s *regionFeedState) getRegionMeta() (uint64, regionspan.ComparableSpan, time.Time, string) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	return s.sri.verID.GetID(), s.sri.span, s.startFeedTime, s.sri.rpcCtx.Addr
+}
+
+type syncRegionFeedStateMap struct {
+	mu     sync.RWMutex
+	states map[uint64]*regionFeedState
+}
+
+func newSyncRegionFeedStateMap() *syncRegionFeedStateMap {
+	return &syncRegionFeedStateMap{
+		mu:     sync.RWMutex{},
+		states: make(map[uint64]*regionFeedState),
+	}
+}
+
+func (m *syncRegionFeedStateMap) setByRequestID(requestID uint64, state *regionFeedState) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.states[requestID] = state
+}
+
+func (m *syncRegionFeedStateMap) takeByRequestID(requestID uint64) (*regionFeedState, bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state, ok := m.states[requestID]
+	if ok {
+		delete(m.states, requestID)
+	}
+	return state, ok
+}
+
+func (m *syncRegionFeedStateMap) takeAll() map[uint64]*regionFeedState {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	state := m.states
+	m.states = make(map[uint64]*regionFeedState)
+	return state
+}
+
+func (m *syncRegionFeedStateMap) setByRegionID(regionID uint64, state *regionFeedState) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.states[regionID] = state
+}
+
+func (m *syncRegionFeedStateMap) getByRegionID(regionID uint64) (*regionFeedState, bool) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	result, ok := m.states[regionID]
+	return result, ok
+}
+
+func (m *syncRegionFeedStateMap) delByRegionID(regionID uint64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	delete(m.states, regionID)
+}
+
+func (m *syncRegionFeedStateMap) len() int {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return len(m.states)
+}
+
+// regionStateManager provides the get/put way like a sync.Map, and it is divided
+// into several buckets to reduce lock contention
+type regionStateManager struct {
+	bucket int
+	states []*syncRegionFeedStateMap
+}
+
+func newRegionStateManager(bucket int) *regionStateManager {
+	if bucket <= 0 {
+		bucket = runtime.NumCPU()
+		if bucket > maxRegionStateBucket {
+			bucket = maxRegionStateBucket
+		}
+		if bucket < minRegionStateBucket {
```

Comment:
It seems that the `maxRegionStateBucket` and `minRegionStateBucket` if statement should be outside the `if bucket <= 0` if statement block, otherwise `bucket` argument can less than this two constraint constant.